### PR TITLE
fix(core): Prevent volatile physical name generation

### PR DESCRIPTION
**Pull request by RomainMuller**
The PhysicalName generation strategy for cross-account/region use-cases
could generate names that are subject to collisions/sniping when the
account and region were not specified; and those names would also have
changed if a stack went from re-locatable to account/region specific,
even if the target environment would not have changed.

The generator will now throw errors in case the account ID or region is
blank or a Token.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by rix0rrr** [comment]
> collisions/sniping

Good point. Given that, a way to add an app-wide salt to the name generation might also be a good idea.

**Comment by RomainMuller** [comment]
@rix0rrr  - yeah the app-wide salt is also in my mind but I decided to NOT do it just yet because this is a little more complex (the salt would be a secret...).

W/R/T the collisions - the _current_ state of things for position-independent stacks is collision-vulnerable indeed. The update here makes the account ID and region *required* part of the name generation, so you couldn't get a generated name for a position-independent stack (and those use-cases typically will involve a determined account and region... because they require cross-account/region workflows).

File: packages/@aws-cdk/cdk/lib/private/physical-name-generator.ts [hunk 1]
```diff
@@ -8,14 +8,14 @@ export function generatePhysicalName(resource: IResource): string {
   const stackPart = new PrefixNamePart(stack.stackName, 25);
   const idPart = new SuffixNamePart(resource.node.uniqueId, 24);
 
-  let region: string | undefined = stack.region;
-  if (Token.isUnresolved(region)) {
-    region = undefined;
+  const region: string = stack.region;
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/private/physical-name-generator.ts]: Technically stack.region/account will never be undefined but that’s ok

**Comment by rix0rrr** [comment]
Yeah I wonder how this would even go wrong.

By the way, we could also make unique hard-name generation a context feature, where we generate a name on-demand and store it it context.

**Comment by rix0rrr** [comment]
Salts would only need to be private for buckets, btw. For other resources, since there's no global namespace there's no chance of sniping, we'd only need unique values.

But yes, to protect against sniping we'd need unique AND secret values.

**Pull request merged**